### ISSUE TITLE
SCons computation of HOST_ARCH is wrong (Visual Studio)

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#2884]
This issue was originally created at: 2012-12-05 14:10:02.
This issue was reported by: `buildsmith`.
buildsmith said at 2012-12-05 14:10:02
>What is broken:
> 
>On 64-bit Windows, support for the recent Visual Studio Express is broken with Python 2.4, 2.5 and 2.7 
> 
>Why could we get along so far?
> 
>Because it always works fine with Python 2.6 (by accident, small defect of Python) and because the issue is more obviously painful with 2012 Express.

buildsmith said at 2012-12-05 14:49:45
> 
>How is it broken:
> 
>SCons code looks for the machine we build on and decides which compilers  to look for based on that. That is a poor idea. For example, on one Win 7 64-bit machine where I want to compile some 32-bit executable, it will try to find a 64-bit built compiler and tell it to build for 32-bit. That search is bound to fail! Recent Express versions  (and also older non-Express versions) do not have any 64-bit built compiler. The sad thing is that there is a 32-bit built compiler on my 64-bit Windows and that will run just fine for my purpose but SCons will not even  look for it.

>As a matter of facts, SCons looks for the wrong vcvars***.bat file. The problem is more obvious with the recent versions of Visual Studio simply because these have more such .bat files to choose from.
> 
> 
>Other remarks:
> 
>Why does it work fine on Python 2.6? Because with Python 2.6 both `platform.machine()` and `PROCESSOR_ARCHITECTURE` wrongly return "x86" on a Win 7 64-bit machine (instead of "amd64").
> 
>The ticket #2849 is an instance of this same problem (and can be closed once the changes below are integrated).

>The reason why some people say about 2849 that it's fixed and some say it's not is the typical confusion between what you build for and what you build with. What is implemented currently in SCons (2.2.0) is a fall back for what you  build for! This is what the later comments claim and it's true. But the original issue of the ticket is the missing fall back for what you build with. And the reporter is right to claim that it is not fixed yet.
> 
>By the way, a fall back for what you build for it's a very poor idea in a production environment. If my customer wants some precise build, say 64-bit binaries, and the build tool changes that behind my back, I'm certainly upset! Notice that the tests will pass OK despite the tool changing what architecture I build for, so the danger to ship out wrong stuff is very real.
> 
> 
>How to fix it:
> 
>Deceptively simple fix. In get_host_target() in file vc.py change the code:
```python
    host_platform = env.get('HOST_ARCH')
    if not host_platform:
        host_platform = platform.machine()
        # TODO(2.5):  the native Python platform.machine() function returns
        # '' on all Python versions before 2.6, after which it also uses
        # PROCESSOR_ARCHITECTURE.
        if not host_platform:
            host_platform = os.environ.get('PROCESSOR_ARCHITECTURE', '')
``` 
>with the code:
```python
    # There is no 64-bit version of Visual Studio and the 64/32 bit-ness of MS Windows or Python on this machine really has no relevance at all
    # Background can be found here:
    #      http://blogs.msdn.com/b/ricom/archive/2009/06/10/visual-studio-why-is-there-no-64-bit-version.aspx
    #      http://news.softpedia.com/news/Don-039-t-Hold-Your-Breath-for-64-bit-Visual-Studio-113943.shtml
    # Even if there are out there some 64-bit command line tools, it's not relevant for SCons, which aims to support the Express editions of Visual Studio as well
    host_platform = "x86"
    if env.get('HOST_ARCH'):
        assert env.get('HOST_ARCH')=="x86", "There no 64-bit version of Visual Studio yet. Only the 'x86' host platform is supported" 
``` 
> 
>How did I test:
> 
>-- MS VS 6, 2003, 2005, 2008, 2010 and 2012, 
>-- Express and non-Express versions for 2005 to 2012
>-- Win 7 64 bits and Win XP 32-bit (SP3 required)
>-- Python 2.5, 2.6 and 2.7 in both 32-bit and 64-bit on the Win 7 machine
>Yes, all those combinations makes up for a whole lot of tests. I also tried separately Windows SDK 6.0A.

buildsmith said at 2012-12-05 15:19:18
>oops, I messed up the "QA contact" field. Now fixed.

bdbaddog said at 2012-12-05 15:52:57
>Don't change the contact info please.

bdbaddog said at 2012-12-06 20:30:08
>Please change the summary to indicate what the problem is. Saying "support 
>problem" is not very descriptive.
> 
>Have you read the  man page, specifically this section:
>"TARGET_ARCH
>Sets the target architecture for Visual Studio compiler (i.e. the arch of the binaries generated by the compiler). If not set, default to $HOST_ARCH, or, if that is unset, to the architecture of the running machine's OS (note that the python build or architecture has no effect). This variable must be passed as an argument to the Environment() constructor; setting it later has no effect. This is currently only used on Windows, but in the future it will be used on other OSes as well. Valid values for Windows are x86, i386 (for 32 bits); amd64, emt64, x86_64 (for 64 bits); and ia64 (Itanium). For example, if you want to compile 64-bit binaries, you would set TARGET_ARCH='x86_64' in your SCons environment.
>"
> 
>As I read your notes on this it sounds like you are looking for `TARGET_ARCH` and don't realize it's available.
> 
>If I'm missunderstanding your reported issue, please reopen.

buildsmith said at 2012-12-07 01:03:22
> 
>Yes, I have to reopen. This ticket is about `HOST_ARCH`. So I changed the title as you suggested. It now reads:  "Scons computation of HOST_ARCH is wrong"
> 
>I have another ticket about the TARGET_ARCH: #2886 and your comment should be copied there in fact. I'll do it. In my paragraph above "By the way..." I alluded to the TARGET_ARCH. I terribly sorry now that I mentioned that. It's not the object
of this ticket.
> 
>The object of this ticket is the fact that what SCons gets in the end in the variable host_platform depends on which version of Python you run.
> 
>I cannot think of any situation where you want the user-observable SCons behavior to depend on the version of Python. 
> 
> 
> 

buildsmith said at 2012-12-07 01:14:33
>From the man page:
> 
>HOST_ARCH
>    Sets the host architecture for Visual Studio compiler. If not set, default to the detected host architecture: note that this may depend on the python you are using.
> 
>So the dependency on the version of Python is acknowledged, true. And the man page does describe accurately what the code does.
> 
>But again, why fool the user and let him think that this is an input  argument at his choice? It is not! We must use the 32-bit built compilers.
> 
> 
> 
> 

buildsmith said at 2012-12-07 01:19:09
> 
>> We must use the 32-bit built compilers.
>I take as granted that the reason is known but may be it's worth repeating. We must use the 32-bit built compilers because this is what Microsoft provides consistently in all versions of Visual Studio and of the SDKs.

buildsmith said this issue blocks #2849 at 2012-12-05 14:49:45.

**Comment by mwichmann** [comment]
Hmmm, this issue is listed as blocking #2849, but that one is already closed. Maybe close this too?

**Comment by bdbaddog** [comment]
> Hmmm, this issue is listed as blocking #2849, but that one is already closed. Maybe close this too?

2849 is about not defaulting TARGET_ARCH to HOST_ARCH, but rather realizing when TARGET_ARCH is not set by the user and then searching for an available TARGET_ARCH.  Which can't happen correctly if HOST_ARCH isn't correct. Thus why this ISSUE is marked as blocking #2849 .

If I'm reading this correctly the but is that depending on the version of Python used on a x64 system HOST_ARCH could be X86 if the python was 32bit?

Maybe due to this: https://blog.differentpla.net/blog/2013/03/10/processor-architew6432/

But SCons logic from 7/20/2009 should handle this and the bug is reported 2012..?

**Comment by bdbaddog** [comment]
From @mwichmann 

PROCESSOR_ARCHITECTURE is x86 inside the 32-bit python
w6432 is AMD64

**Comment by bdbaddog** [comment]
Hmm.. this wording is still in the manpage though:

> If not set, default to the detected host architecture: note that this may depend on the python you are using.

**Comment by bdbaddog** [comment]
So this needs updating minimally: https://github.com/SCons/scons/blob/master/src/engine/SCons/Tool/msvc.xml#L402

**Comment by mwichmann** [comment]
Anybody have clues what the current state is?  A few things have changed as I recall in the handling of *ARCH, the difference between pre-2.7 and 2.7 Python is moot now the floor version is 3.6, VS Express 2012 is a decade old and unsupported, etc.

**Comment by bdbaddog** [comment]
I created this test SConstruct

```
env=Environment()

print("HOST_ARCH: %s"%env['HOST_ARCH'])
```

Then I ran it on 32 and 64 bit python 3.10.

```
PS C:\Users\Bill\devel\scons\git\scons-bugswork\2884> py -3.10-32 ..\..\as_scons\scripts\scons.py
scons: Reading SConscript files ...
HOST_ARCH: x86_64
scons: done reading SConscript files.
scons: Building targets ...
scons: `.' is up to date.
scons: done building targets.
PS C:\Users\Bill\devel\scons\git\scons-bugswork\2884> py -3.10 ..\..\as_scons\scripts\scons.py
scons: Reading SConscript files ...
HOST_ARCH: x86_64
scons: done reading SConscript files.
scons: Building targets ...
scons: `.' is up to date.
scons: done building targets.
```

So, if I understand this bug correctly, the current bug is resolved.
@mwichmann thoughts?

**Comment by jcbrill** [comment]
*TL;DR - I believe that this issue can be closed.  Read on for the gory details...*

Results of testing with Python 2.7 in a 64-bit windows vm with scons 2.2.0 and 2.4.0 reveal that there was an "issue" in scons 2.2.0 when the ```TARGET_ARCH``` was not specified that was resolved by 2.4.0.

The title of the bug report is somewhat misleading as the ```HOST_ARCH``` calculation is technically correct.

As noted earlier, the ```HOST_ARCH``` returned in Python 2.7 is different than that returned in python 2.6 and earlier.  The same holds for Python 3.2 and later versus Python 3.1 and earlier.

The environment was initialized in scons 2.2.0 environment.py as:
```
self._dict['HOST_OS']      = self._dict.get('HOST_OS',None)
self._dict['HOST_ARCH']    = self._dict.get('HOST_ARCH',None)
        
# Now set defaults for TARGET_{OS|ARCH}
self._dict['TARGET_OS']      = self._dict.get('HOST_OS',None)
self._dict['TARGET_ARCH']    = self._dict.get('HOST_ARCH',None)
```

Given an empty environment:
```
env = Environment()
```

There is a difference in behavior between python 2.6 (and earlier) and python 2.7 when the ```TARGET_ARCH``` is not user-defined:
* In python 2.6: ```TARGET_ARCH = 'x86'```
* In python 2.7: ```TARGET_ARCH = 'x86_64```

Effectively, ```TARGET_ARCH``` is forced to a non-None value.  This causes an issue in vc.py.  In the vc.py code, if ```TARGET_ARCH``` is not None it is assumed to be a user request.


For python 2.7, vc.py believes the user requested a ```TARGET_ARCH``` of ```x86_64```.  As written at the time, the vc.py code does not attempt the ```x86``` targets: 
```
    # If the user hasn't specifically requested a TARGET_ARCH, and
    # The TARGET_ARCH is amd64 then also try 32 bits if there are no viable
    # 64 bit tools installed
    try_target_archs = [target_platform]
    if not req_target_platform and target_platform in ('amd64','x86_64'):
        try_target_archs.append('x86')
```

This is a problem for the Express versions.

For python 2.6, vc.py believes the user requested a ```TARGET_ARCH``` of ```x86```. This would work regardless of full or express version.

Given an environment where the ```TARGET_ARCH``` is specified:
```
env = Environment(TARGET_ARCH='x86')
```

This works as expected as the ```TARGET_ARCH``` initialized value is over-written with the user-defined value.

In scons 2.4.0, the environment initialization appears as it does today:
```
self._dict['HOST_OS']      = self._dict.get('HOST_OS',None)
self._dict['HOST_ARCH']    = self._dict.get('HOST_ARCH',None)

# Now set defaults for TARGET_{OS|ARCH}
self._dict['TARGET_OS']      = self._dict.get('TARGET_OS',None)
self._dict['TARGET_ARCH']    = self._dict.get('TARGET_ARCH',None)
```

In scons 2.4.0, an empty environment works as expected with both python 2.6 and python 2.7.  

Defining a environment with an explicit ```TARGET_ARCH``` request works with both python 2.6 and python 2.7 for scons 2.2.0 and 2.4.0.

In the current version of SCons and *only* VS2008 Express installed, #2697 still applies due to issues with the initial detection.

**Comment by bdbaddog** [comment]
Great! Thanks for digging into this!
Closing.
Yes I initially added the TARGET_* logic, and then fixed it as above.
For many settings it is good to be able to tell if the values are defaults or user requested.
